User defined indexing operators: updated

Octachron opened the pull request:
This PR updates the user-defined-indexing-operators branch past the last modifications of the bootstrapping compiler. Due to the necessity to preserve the indexing operators syntax and primitives within the bootstrapping compiler, the merge history is quite messy. An alternative path might be to recreate the branch by cherry-picking the relevant commits.

Octachron committed:
Indexing submodule for bigarray

The objective of this commit is to introduce a short notation for
bringing in scope the bigarray index operators and only them. For
that purpose, the bigarray index operators are regrouped in a single
submodule. This submodule is also included inside the global bigarray
module to preserve compatibility and ease of use of the bigarray
module.

Octachron committed:
Bigarray index operators deprecated warning

With the simplification of index operators, the expressions a.{..} are
no longer automatically resolved to Bigarray.Array[n].[g|s]et. To use
these operators, it is now necessary to bring them in scope, for
instance by opening either the Bigarray or Bigarray.Operators module.
To ease the transition period, this patch add an hack in
`typing/typetexp.ml` to catch the cases where the index operators
`.{}/.{,}..` are used without being bound in the current scope
and tranlate then to Bigarray.(..)  with a deprecated warning.

gasche committed:
Merge pull request #334 from Octachron/ba_indexing

User defined indexing operators: operator submodule for bigarray

Octachron committed:
Manual: Document bigarray compatibility warning

This commit update the documentation on the compatibility problems
between the deprecated bigarray specific syntax extension and the new
user-defined index operators extension. In particular, this commit
describes the new deprecated warning for implicit use of the
`Bigarray(.{...})` operators and states that this warning might be
turned into an error in the undetermined futures. This commits also
amend the documentation to mention the new `Bigarray.Operators`
submodule when useful.

Octachron committed:
Manual: Fix user-defined index operators

Change the name from customizable to user-defined index operators and
fix the alignment of the latex tables for better readability.

Octachron committed:
Fix GPR#69 changes: bigarray compatibility section

gasche committed:
Merge pull request #339 from Octachron/bigarray_deprecated_warning

User-defined indexing operators: bigarray deprecated warning

mshinwell committed:
Flambda manual chapter

gasche committed:
GPR#387, MPR#7172: document ./configure -no-native-compiler

gasche committed:
./configure: set natdynlink=false on -no-native-compiler

In PR#7172, Daniel Bünzli remarks that even if an user does only "make
world install" only, intentionally not installing the native code
compiler, the configuration variables exposed through
$(ocamlc -where)/Makefile.config will be misleading as they indicate
the configuration-time prediction that this installation has a native
compiler and support native dynlink.

The -no-native-compiler switch allows to avoid this issue by
intentionally indicating at configure-time that no native compiler
shall be built. However, its current implementation forgot to also
explicitly disable native dynlinking. Changing this fixes the
Makefile.config information and has no impact on the compiler build
process (a build with ARCH=none will never hit the {all,install}opt of
dynlink/Makefile that inspect NATDYNLINK).

On regression risk: it is easy to check in ./configure that this
change only affects configuration when the -no-native-compiler option
is passed, which did not exist before (late in) 4.03+dev.

alainfrisch committed:
Avoid escaping inline record type through type abbreviation [type 'a t = t].

damiendoligez committed:
PR#7180: Gc.major_slice throws off GC work calculations

damiendoligez committed:
documentation for PR#6808: the parsing of OCAMLRUNPARAM is too lax

damiendoligez committed:
PR#6668: Missing "type-constraint" in grammar for "mod-constraint"

alainfrisch committed:
GPR#523: special support for camlp4/camlp5 'explicit arity'.

alainfrisch committed:
Fix linking order.

damiendoligez committed:
PR#7180, second bug:

caml_alloc_custom ignores the `mem` and `max` parameters when allocating
in the minor heap.

gasche committed:
MPR#7196: "let open" is not correctly pretty-printed to the left of a ';'

lpw25 committed:
Fix documentation comments on let bindings

lpw25 committed:
Remove ;; from stdlib mli files

lpw25 committed:
Remove ;; from stdlib ml files

lpw25 committed:
Turn on warning 50 for the standard library

lpw25 committed:
Turn on warning 50 for otherlibs

garrigue committed:
Fix PR#7199

garrigue committed:
Fix PR#7181

mshinwell committed:
Attend to final Flambda CR comments for 4.03 release

chambart committed:
Make Cmmgen understand constant bools

lpw25 committed:
Merge pull request #528 from lpw25/remove-stdlib-semisemi

Remove `;;` from the standard library

mshinwell committed:
Fix spelling mistake in comment

mshinwell committed:
Merge pull request #503 from mshinwell/flambda_doc

Flambda manual chapter

damiendoligez committed:
last commit before tagging 4.03.0+beta2

damiendoligez committed:
change VERSION for 4.03.0+beta2

damiendoligez committed:
increment version number after tagging 4.03.0+beta2

garrigue committed:
Fix PR#7207

Octachron committed:
Merge commit 'c6b5a28d77309931df0' into user-defined-indexing-operators

Ignore user-defined indexing operators revert

Commit:
PR#7209: do not run at_exit handlers on exec failure

Octachron committed:
Merge commit 4.03-part 1 into user-defined-indexing-operators

Octachron committed:
Merge 4.03 - part 2 into user-defined-indexing-operators

Octachron committed:
Merge 4.03-part 3 into user-defined-indexing-operators

Octachron committed:
Merge 4.03-part 4 into user-defined-indexing-operators

Octachron committed:
Merge 4.03-part 5 into user-defined-indexing-operators

Octachron committed:
Merge 4.03-part 6 into user-defined-indexing-operators

Octachron committed:
Merge 4.03-part 7 into user-defined-indexing-operators

Octachron committed:
Merge 4.03-part 8 into user-defined-indexing-operators

damiendoligez committed:
complement to commit c95ccf7e9b3df76fb159113c36312a4e9914f627 (PR#7180, second bug)

damiendoligez committed:
Merge GPR#474 into trunk:
  Minor tidying of middle-end

damiendoligez committed:
Merge pull request #532 from diml/fix-mantis-7209

PR#7209: do not run at_exit handlers on exec failure

damiendoligez committed:
Merge pull request #529 from lpw25/warning-50-stdlib

Warning 50 stdlib

damiendoligez committed:
Merge pull request #526 from lpw25/fix-let-binding-comments

Fix documentation comments on let bindings

Octachron committed:
Merge '4.03' into user-defined-indexing-operators

gasche commented:
Apologies for not maintaining the branch as I said I would. I would like to do the rebase by cherry-picking, but I have no time for this currently. I'll look at this again after the release -- feel free to ping me.

Octachron commented:
Note that I also have branches with cherry-picked version of the changes, if you prefer:
rebased on 4.03: https://github.com/Octachron/ocaml/tree/index_operators_updated
rebased on trunk: https://github.com/Octachron/ocaml/tree/index_operators_+

gasche commented:
So I just changed [ocaml/user-defined-indexing-operators](https://github.com/ocaml/ocaml/tree/user-defined-indexing-operators) to point to your `index_operators_+` branch. Thanks!
